### DISCUSSION TITLE
Json properties not updated when set by index

### DISCUSSION THREAD

**Discussion by snigdha920** [mikro-orm/mikro-orm#4225] [Q&A]
We have an issue where MikroORM is sometimes not saving changes to the database. 

I have a billing entity:

```typescript
@Entity({ customRepository: () => BillingRepository })
export class Billing extends AbstractEntity {
  public [EntityRepositoryType]?: BillingRepository;
  
  // creditsHistory maps an invoice id to an amount of credits that were spent
  @Property({ type: 'json' })
  public creditsHistory: { [key: string]: number } = {};

  // more fields
}
```

and when we:

1. Fetch the entity from it's custom repository:

```typescript
const billingRepository = this.billingRepository;
const billing = await billingRepository.findByCustomer(customerId);
```

2. Make some changes

```
billing.creditsHistory[invoiceId] = creditsToSpend;
```

3. Run `persistAndFlush` (I understand that `persist` is for the first time when we are creating an entity, but this has been working since we were on v4)

```
await billingRepository.persistAndFlush(billing);
```

The updates are not being saved in the database sometimes. There's also no error logs, just the changes are not saved. We're on MongoDB, here's all our version numbers:

```typescript
    "@mikro-orm/core": "5.6.8",
    "@mikro-orm/mongodb": "5.6.8",
    "@mikro-orm/nestjs": "5.1.8"
```

and our Mikro config: 

```typescript
      type: 'mongo',
      clientUrl,
      dbName,
      debug: this.get<boolean>('mongoDebug'),
      ensureIndexes: !this.isTestRun(),
      implicitTransactions: false,
      validate: true,
      strict: true,
      populateAfterFlush: true,
      autoLoadEntities: false, // does not work for tests and workers, so we do it manually everywhere
      registerRequestContext: false, // we fork the em explicitly in the app startup
      findOneOrFailHandler: (entityName, where) => {
        return new NotFoundException(`${entityName} where ${JSON.stringify(where)})} does not exist`);
      },
```

I wrote a test to reproduce the issue, but the test passed and I couldn't reproduce it there.

We've been using MikroORM since v4 and we started seeing such issues around v5.6.8. These issues are flaky and they happen sometimes. 

Do you have a hunch as to what may be causing this? 

This is causing us to lose a lot on money in production, and without a way to consistently reproduce it, they are very hard to debug as well.

**Comment by B4nan** [comment]
Can you try latest dev version? I was doing some fixes for json properties in there.

If you can reproduce locally in your app, I can guide you where to put console logs to get more insights on what is happening. But I am on vacation now and dont want to do it from cell phone :) Its mostly in the UoW and changesetcomputer - there is a method that calculates the difference, you could log the original data and current data to see what it tries to compare. That way it should be clear if the problem comes from diffing or rather snapshotting (saving the entity state for later diffing).

**Reply by snigdha920** [reply]
Thanks for replying on your holiday! 

I'm on version `5.6.17-dev.9`, and I am now able to reproduce the issue with this test:

```typescript
  it('should update the creditsHistory', async () => {
    const billing = new Billing(workspace, stripeCustomerId, 0);
    await billingRepository.persistAndFlush(billing);

    em.clear();

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const billing2 = (await billingRepository.findByCustomer(stripeCustomerId))!;

    billing2.credits = 10000;
    billing2.creditsHistory['testInvoiceId'] = 24000;

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const originalBilling2 = wrap(billing2, true).__originalEntityData!;

    expect(originalBilling2.credits).toEqual(0);
    expect(originalBilling2.creditsHistory).toEqual({});
  });
```

The test fails with the error:

```
    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 3

    - Object {}
    + Object {
    +   "testInvoiceId": 24000,
    + }
```

So the `credits` which is of type `number` is updated correctly, but not the `creditsHistory` of type `json`.

**Reply by snigdha920** [reply]
Wrote a quick test in the mikro orm repo and it works as expected: I'm really at a loss for where I can find and fix this issue, do you have any ideas?

```typescript
  test('json prop on original entity', async () => {
    const author = new Author('Jon Snow', 'snow@wall.st');
    await orm.em.persistAndFlush(author);

    orm.em.clear();
    const repo = orm.em.getRepository(Author);
    const author2 = (await repo.findOne(author.id))!;

    // eslint-disable-next-line dot-notation
    author2.creditsHistory['foo'] = 1;
    expect(wrap(author2, true).__originalEntityData?.creditsHistory).toMatchObject({});
  });
```

**Reply by snigdha920** [reply]
> That way it should be clear if the problem comes from diffing or rather snapshotting (saving the entity state for later diffing).

Edit: Here's the log at `EntityComparator.diffEntities(entityName, a, b)`. So it's calculating the diff incorrectly.

```typescript
at EntityComparator.diffEntities (../../../../node_modules/.pnpm/@mikro-orm+core@5.6.17-dev.9_dcri46wahboejbzhwbxuikaayi/node_modules/@mikro-orm/core/utils/EntityComparator.js:26:17)

  console.log
    a is  {
      _id: ObjectId('643c3479b1a40efb6880e1ef'),
      createdAt: ISODate('2023-04-16T17:46:33.430Z'),
      updatedAt: ISODate('2023-04-16T17:46:33.473Z'),
      workspace: ObjectId('643c3479b1a40efb6880e1ee'),
      credits: 10000,
      upcoming: 0,
      billingStatus: 'active',
      creditsHistory: { testInvoiceId: 24000 },
      stripeCustomerId: 'cus_K2mP2QQs3Ft4vi',
      stripeSubscriptions: [],
      stripeSubscriptionMap: {},
      stripeSubscriptionItemMap: {},
      creditWarningSent: false
    }

      at EntityComparator.diffEntities (../../../../node_modules/.pnpm/@mikro-orm+core@5.6.17-dev.9_dcri46wahboejbzhwbxuikaayi/node_modules/@mikro-orm/core/utils/EntityComparator.js:27:17)

  console.log
    b is  {
      _id: ObjectId('643c3479b1a40efb6880e1ef'),
      createdAt: ISODate('2023-04-16T17:46:33.430Z'),
      updatedAt: ISODate('2023-04-16T17:46:33.473Z'),
      workspace: ObjectId('643c3479b1a40efb6880e1ee'),
      credits: 10000,
      upcoming: 0,
      billingStatus: 'active',
      creditsHistory: {},
      stripeCustomerId: 'cus_K2mP2QQs3Ft4vi',
      stripeSubscriptions: [],
      stripeSubscriptionMap: {},
      stripeSubscriptionItemMap: {},
      creditWarningSent: false
    }

      at EntityComparator.diffEntities (../../../../node_modules/.pnpm/@mikro-orm+core@5.6.17-dev.9_dcri46wahboejbzhwbxuikaayi/node_modules/@mikro-orm/core/utils/EntityComparator.js:28:17)

  console.log
    diff is { creditsHistory: {} }
```

**Reply by snigdha920** [reply]
Wow, in the end I get the test to work if I just do:

```typescript
billing.creditsHistory = {testInvoiceId: 24000}
```

instead of:

```typescript
billing2.creditsHistory['testInvoiceId'] = 24000;
```

I added the `dot-notation` eslint rule as a "fix". If you have any other ideas on why this is happening, much appreciated.

**Reply by B4nan** [reply]
Well, those two lines are not doing the same - first one is setting a new object with one property, the other is setting a value of existing object. There is no runtime difference for "bracket" vs "dot" notation, they are doing exactly the same thing.

**Reply by snigdha920** [reply]
Oh yes indeed. 

> Well, those two lines are not doing the same - first one is setting a new object with one property, the other is setting a value of existing object.

The issue then is for `json` and `array` types, we need to always re-assign, and updating the existing object/array (eg. `array.push`, `object[key] = value`) doesn't save the changes in the database. That's because the `originalEntity` itself is getting updated when we update the object/array, when it should remain the same "original" value. 

I've tried this on `5.6.17-dev.10` and `5.6.16` too now. Can reproduce it on both the versions.

**Reply by B4nan** [reply]
Hmm so sounds like we are missing some `clone` call, the originalEntityData should never contain references to the same values as the actual entity.

The method that saves snapshot is `UoW.registerManaged`, you could put some logs in there, including `new Error().stack` to see where it is coming from. The snapshot itself is generated via `EntityComparator.prepareEntity`, but this is probably coming from other code paths, I think snapshotting works as expected. Maybe `EntityFactory.mergeData`.
